---
title: "tfbsDB: Store and plot TFBS sites from HOCOMOCO v11 scored by FIMO"
author: "[Gabriel Hoffman](http://gabrielhoffman.github.io)"
date: "`r format(Sys.Date(), '%m/%d/%Y')`"
abstract: > 
  Store and plot TFBS sites from HOCOMOCO v11 scored by FIMO
  tfbsDB 
  v`r packageVersion("tfbsDB")`<br> <br> 

output:
  rmarkdown::html_document:
    highlight: pygments
    toc: false
    toc_depth: 3
    fig_width: 8
vignette: >
  %\VignetteIndexEntry{tfbsDB: Store and plot TFBS sites from HOCOMOCO v11 scored by FIMO}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
  %\usepackage[utf8]{inputenc}
---

<!---
cd /Users/gabrielhoffman/workspace/repos/
R
library('knitr')
rmarkdown::render('tfbsDB/vignettes/tfbsDB.Rmd')

# documentation
LOCATION=/Users/gabrielhoffman/workspace/repos/tfbsDB
rm -f ${LOCATION}/man/*.Rd
Rscript -e "library(roxygen2); roxygenise('${LOCATION}')"; 

library(roxygen2); 
roxygenise('/Users/gabrielhoffman/workspace/repos/tfbsDB')

R CMD build tfbsDB

R CMD INSTALL tfbsDB

R CMD check tfbsDB

R CMD BiocCheck tfbsDB

--->

```{r setup, echo=FALSE, results="hide"}
knitr::opts_chunk$set(tidy=FALSE, cache=TRUE,
                      dev="png",
                      message=FALSE, error=FALSE, warning=TRUE)
options(width=100)
```	

tfbsDB includes functions to read TFBS from bgzip'd and tabix'd BED file, and plot using ggbio.
```{r tfbs_Example}
library(GenomicRanges)
library(tfbsDB)

# range of TFBS to be loaded from file
grQuery = GRanges("chr1", IRanges(10280, 10410))

# example data from package
bedFile = file.path(system.file("data", package = "tfbsDB"), "motifs_fimo_exmaple.bed.gz")

# read from file
gr = readTFBSdb( grQuery, bedFile )

# show TFBS locations
plotTFBSdb( gr, xlim=c(10280, 10410), colorByP=FALSE)

# color locations by p-value of motif match
plotTFBSdb( gr, xlim=c(10280, 10410), colorByP=TRUE)
```

Plot genes and TFBS
```{r tfbs_genes, fig.width=6}
library(ggbio)
library(tfbsDB)
library(EnsDb.Hsapiens.v75)
edb <- EnsDb.Hsapiens.v75

# chr17:41,323,246-41,363,707
wh <- GRanges("chr17", IRanges(41323246,41333246))
xlim = c(start(wh), end(wh))

# plot Genes
wh2 <- wh
seqlevelsStyle(wh2) <- "NCBI"
p.txdb = autoplot(edb, which=wh2, names.expr = "gene_name") + theme_bw(20) 

# plot TFBS
bedFile = '~/data/tfbsDB/motifs_fimo_search_small.bed.gz'
gr = readTFBSdb( wh, bedFile )
figTFBS = plotTFBSdb( gr[1:10], xlim, colorByP=FALSE, tf_text_size=3) 

# Combine
tracks( Genes = p.txdb, TFBS = figTFBS, label.bg.fill="navy", label.text.color="white", 
  heights=c( 1, .4), padding = unit(-1.9, "lines"), label.width = unit(1.5, "lines"), 
  xlab.height = unit(1, "lines"), xlim=xlim) + scale_x_sequnit("bp")
```



# Session info
```{r sessionInfo}
sessionInfo()
```


